### DISCUSSION TITLE
In 2.0, how do I properly use scoped_session() to get typing hints?

### DISCUSSION THREAD

**Discussion by wu-clan** [sqlalchemy/sqlalchemy#8728] [Usage Questions] [answered]
I have an existing example :
```python
db_session = sessionmaker(bind=engine, autoflush=False)

def get_db():
    session = scoped_session(db_session)
    return session
```

when:
```python
def get_first():
    with get_db().begin() as db:
        return db.xxx
```
![image](https://user-images.githubusercontent.com/52145145/198297808-d7678fdb-c33a-44a8-8a65-fb79d1f05fbc.png)

when:
```python
def get_first():
    with get_db().begin() as db:
        return db.session.xxx  # <- different
```
![image](https://user-images.githubusercontent.com/52145145/198300546-d92a7ac5-fe25-4755-bf20-e8bd01dc9b07.png)

`scoped_session().begin()` returns SessionTransaction, it has an attribute value `session: Session`, so can get typed hints via `db.session.xxx`, but it's weird to use `db.session.xxx` for hints

**Comment by zzzeek** [comment, marked as answer]
overall, you would not do this:

```py

def get_db():
    session = scoped_session(db_session)
    return session
```

scoped_session() is intended to be used as a global variable.  I would not advise using it.    

the pattern we document for Session by itself is at https://docs.sqlalchemy.org/en/14/orm/session_basics.html#framing-out-a-begin-commit-rollback-block

```py
with get_db() as session, session.begin():
     # ...
```

however, if you use sessionmaker, begin() gives you the session, which is analogous to create_engine():

```py
Session = sessionmaker()

with Session.begin() as session:
    session.add(some_object)
    session.add(some_other_object)
```



short answer:   dont use scoped_session

**Comment by wu-clan** [comment]
Thanks, I'm migrating from 1.4 to 2.0 and started using session.begin() globally, it works fine, out of excellent documentation, for scoped_session I simply interpreted it as thread safe, so trying it out, if it's not recommended I'll drop it

**Reply by zzzeek** [reply]
right there is nothing inherently wrong with scoped_session except that people don't really understand it.    saying it's "thread safe" is not really descriptive of what it actually means, and if you have a pattern where new sessions are being created within functions in any case, you dont need the Session to be "thread safe".    

so in general that's why i recommend not using it, because it's not obvious to understand what it is.

**Reply by wu-clan** [reply]
Thanks for the clarification, I think I used session.begin() enough
